### PR TITLE
Failing test for DI's site call factory cache

### PR DESCRIPTION
**Pull request by allantargino**
failing test for issue #57333

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs [hunk 1]
```diff
@@ -361,6 +388,13 @@ public void Dispose()
             }
         }
 
+        private static T GetServiceFromCallSiteFactory<T>(ServiceProvider provider)
+        {
+            var callSite = provider.CallSiteFactory.GetCallSite(typeof(T), new CallSiteChain());
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs:393]: Can we write this test in terms of the public API? I don't think we should be using `internal` APIs unless we absolutely must.

Comment (allantargino) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs]: Done :) The test even became simpler. Thanks

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs [hunk 2]
```diff
@@ -271,6 +272,31 @@ private void NoServiceCallsite_DefaultValueNull_DoesNotThrow(ServiceProviderMode
...
 
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/57333")]
+        public void CallSiteFactoryResolvesIEnumerableOfOpenGenericServiceAfterResolvingClosedImplementation()
+        {
+            IServiceCollection descriptors = new ServiceCollection();
+            descriptors.Add(ServiceDescriptor.Scoped(typeof(IFakeOpenGenericService<int>), typeof(FakeIntService)));
+            descriptors.Add(ServiceDescriptor.Scoped(typeof(IFakeOpenGenericService<>), typeof(FakeOpenGenericService<>)));
+
+            var provider = new ServiceProvider(descriptors, ServiceProviderOptions.Default);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs:283]:
```suggestion
            ServiceProvider provider = descriptors.BuildServiceProvider();
```

**Review by eerhardt — APPROVED** [review]
LGTM, except for one tiny correction to use public API.

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/CallSiteTests.cs]: Committing to get CI to run, so this change can be merged.

**Pull request merged**